### ISSUE TITLE
openetelemetry exporter grpc protocol option require setting channel config on grpcbox

### ISSUE THREAD

**Issue by calvin-kargo** [open-telemetry/opentelemetry-erlang#370]
related to PR https://github.com/open-telemetry/opentelemetry-erlang/pull/366

without setting channel config on grpcbox, opentelemetry_exporter fail with these error
> [info][2022-02-25 09:53:57.920][module=otel_exporter function=init/1 ] OTLP exporter successfully initialized
[info][2022-02-25 09:53:57.936][module=opentelemetry_exporter function=export/3 ] OTLP grpc export failed with error: :undefined_channel

Setting config grpcbox like below help fix the error :
```
config :grpcbox, client: %{
  channels: [
    {:opentelemetry_exporter, [{:http, "localhost", "51478"}], %{}}
  ]
}
```

however, i just noticed, that adding config grpcbox like that would make a new error popup (unable to start grpc channel on opentelemetry exporter)
> [warning][2022-02-25 10:03:07.805][module=opentelemetry_exporter function=init/1 ] unable to start grpc channel for exporting and falling back to http_protobuf protocol. reason={:error, {:already_started, #PID<0.6178.0>}}
[info][2022-02-25 10:03:07.805][module=otel_exporter function=init/1 ] OTLP exporter successfully initialized
[info][2022-02-25 10:03:07.872][module=opentelemetry_exporter function=export/3 ] client error exporting spans {:failed_connect,
 [{:to_address, {'localhost', 4137}}, {:inet, [:inet], :econnrefused}]}

on elixir 1.13.3 OTP 24, with library version:
```
:opentelemetry, "1.0.2"
:opentelemetry_api, "1.0.2"
:opentelemetry_ecto, "1.0.0-rc.5"
:opentelemetry_exporter, "1.0.2"
:opentelemetry_phoenix, "1.0.0-rc.7"
:opentelemetry_telemetry, "1.0.0-beta.7"
:grpcbox, "0.14.0"
```

also, opentelemetry & opentelemetry_exporter is started via releases config, 
```
def project() do
   [
      releases: [
        service: [
          applications: [opentelemetry_exporter: :permanent, opentelemetry: :temporary]
        ]
      ]
end
```

**Comment by calvin-kargo** [comment]
This is just a hunch, but trying to make sense of the error by looking at the codebase once more on opentelemetry_exporter and grpcbox, here's some consideration:
- opentelemetry_exporter should automatically start channel grpcbox when `init` is called , this is indicated by `reason={:error, {:already_started, #PID<0.6178.0>}}` in which starting channel failed because channel already started if using config
- however, still the problem of undefined channel, which to my amateur eye indicate two possibility (sorry i am not well versed in erlang): either autocreated channel is not registered in `gproc_pool` (which make when trying to find on `pick` on `grpcbox_channel` return `undefined_channel`, or somehow autocreated channel die after it's init

**Comment by calvin-kargo** [comment]
cc @tsloughter

**Comment by calvin-kargo** [comment]
I have done some more test, i tried to run below command after server started (via `iex -S mix phx.server`)
``` elixir
:grpcbox_channel.start_link(:opentelemetry_exporter, [], %{})
:gproc_pool.pick_worker(:opentelemetry_exporter)
```

Here's the result:
Without config grpcbox:
```
[info][2022-02-25 10:48:38.208][module=otel_exporter function=init/1 ] OTLP exporter successfully initialized
 
[info][2022-02-25 10:48:38.244][module=opentelemetry_exporter function=export/3 ] OTLP grpc export failed with error: :undefined_channel
 
iex(4)> :grpcbox_channel.start_link(:opentelemetry_exporter, [], %{})
{:ok, #PID<0.967.0>}
iex(6)> :gproc_pool.pick_worker(:opentelemetry_exporter)
false
```

With config grpcbox:
```
[warning][2022-02-25 10:51:41.163][module=opentelemetry_exporter function=init/1 ] unable to start grpc channel for exporting and falling back to http_protobuf protocol. reason={:error, {:already_started, #PID<0.745.0>}}
 
[info][2022-02-25 10:51:41.163][module=otel_exporter function=init/1 ] OTLP exporter successfully initialized
 
[info][2022-02-25 10:51:41.238][module=opentelemetry_exporter function=export/3 ] client error exporting spans {:failed_connect,
 [{:to_address, {'localhost', 4137}}, {:inet, [:inet], :econnrefused}]}
 
nil
iex(3)> :grpcbox_channel.start_link(:opentelemetry_exporter, [], %{})
{:error, {:already_started, #PID<0.745.0>}}
iex(4)> :gproc_pool.pick_worker(:opentelemetry_exporter)
false
```

So it seems that the auto-created grpcbox_channel died after creation, though for both config or without config `:gproc_pool.pick_worker(:opentelemetry_exporter)` return false

**Comment by tsloughter** [comment]
Thanks for the details. What is your config for opentelemetry and opentelemetry_exporter?

**Comment by calvin-kargo** [comment]
based on elixir application, the only config of opentelemetry on `releases.exs` is
```
# OpenTelemetry Exporter is set on default config
# In case customization needed, using Env Var to set option is preferable, see https://hexdocs.pm/opentelemetry_exporter/readme.html#os-environment
config :opentelemetry_exporter,
  otlp_protocol: :grpc,
  otlp_endpoint: "http://localhost:4137"
```

**Comment by jowrjowr** [comment]
In the interest of "making this work while this issue gets sorted", using the oltp_protocol of ":http_protobuf" is sufficient to get my traces flowing again.

**Comment by tsloughter** [comment]
Oh crap @calvin-kargo that config has a typo fro the docs that only got fixed recently :(

The port should be 4317.

**Comment by tsloughter** [comment]
Tho if I try to reproduce by having a collector listening on port 4317 and configuring the exporter to use 4137 I don't get an undefined channel error, I get a connection error:

```
(otel_getting_started@rosa)3> 2022-03-07T17:49:27.075433-07:00 info: OTLP grpc export failed with error: econnrefused
```

So maybe that isn't it.

**Comment by calvin-kargo** [comment]
@jowrjowr right, using `http_protobuf` bypass this issue (which i eventually used also).

**Comment by calvin-kargo** [comment]
@tsloughter for reproducibility of the issue, i made a simple phoenix application with dockerfile on https://github.com/calvin-kargo/test_grpc_opentelemetry

though the error is slightly different, it first start with:
`[info] OTLP grpc export failed with error: :no_endpoints`
then become
`[info] OTLP grpc export failed with error: :undefined_channel`

**Comment by tsloughter** [comment]
Thanks!

It seems it is a bug with how `otlp_endpoint` configuration is handled because when I added `ENV OTEL_EXPORTER_OTLP_TRACES_ENDPOINT="http://localhost:4317"` it works.

Will dig into it more today.

**Comment by tsloughter** [comment]
aaaaaaaaaaaaaaaagh :)

I found the issue, binaries! I couldn't reproduce because I kept trying in Erlang. And it works with the OS environment variable because that comes in as an Erlang charlist, not a binary.

There should be code that converts the binary strings of Elixir to Erlang charlists for use in like httpc and grpcbox, but that isn't working apparently.

I'll get this fixed up and some tests added.
